### PR TITLE
Add evolution method for Gimmighoul, EVO_LEVEL_ITEM_COUNT_999

### PR DESCRIPTION
**Pull request by kittenchilly**
## Description
Adds the evo method for Gimmighoul; if you have 999 Gimmighoul coins in the bag and level up, your Gimmighoul evolves into Gholdengo and removes the 999 coins. You can set EVO_LEVEL_ITEM_COUNT_999 to require 999 of any item for custom evolutions as well. 
This evolution method is a bit unique, as it prevents you from stopping it since, if you could, it'd scam you out of all your coins. Because of this, I had to add a new evo mode alongside it, EVO_MODE_CANT_STOP, which is just EVO_MODE_NORMAL, but it doesn't let you stop the evolution. This might be useful for more stuff later on as well.

## Feature(s) this PR does NOT handle:
Config option perhaps to allow stopping the evo but refunding the coins, which would be pretty hard to do (and is not how it works officially)

## **Discord contact info**
kittenchilly

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
`EVO_LEVEL_COLLECT_999` maybe a different name, something like `EVO_LEVEL_ITEM_COUNT_999`?

**Comment by kittenchilly** [comment]
> `EVO_LEVEL_COLLECT_999` maybe a different name, something like `EVO_LEVEL_ITEM_COUNT_999`?

I followed what it was called in SV's data (LevelUpCollect999) but I guess we can change it.

**Comment by AlexOn1ine** [comment]
> > `EVO_LEVEL_COLLECT_999` maybe a different name, something like `EVO_LEVEL_ITEM_COUNT_999`?
> 
> I followed what it was called in SV's data (LevelUpCollect999) but I guess we can change it.

I feel like Collect999 doesn't say much so I would prefer something different. I'm pretty sure we deviated on a lot of labels so this shouldn't be a problem.

**Comment by LOuroboros** [comment]
`EVO_LEVEL_ITEM_COUNT_999` sounds good to me too. It makes it clear that the evo method acts if the Player has 999 copeies of the specified item to me.

File: src/party_menu.c [hunk 1]
```diff
@@ -5753,10 +5767,25 @@ static void PartyMenuTryEvolution(u8 taskId)
...
+            DestroyTask(taskId);
+        }
         else
-            gTasks[taskId].func = Task_ClosePartyMenuAfterText;
+        {
+            if (gPartyMenu.menuType == PARTY_MENU_TYPE_FIELD && CheckBagHasItem(gSpecialVar_ItemId, 1))
+                gTasks[taskId].func = Task_ReturnToChooseMonAfterText;
+            else
+                gTasks[taskId].func = Task_ClosePartyMenuAfterText;
+        }
```
Comment (AlexOn1ine) [inline src/party_menu.c:5788]: Could you explain why this change was needed?

Comment (kittenchilly) [reply src/party_menu.c:5788]: Since EVO_MODE_CANT_STOP is the same as EVO_MODE_NORMAL but you cant stop the evolution, I had to edit every place that called BeginEvolutionScene or EvolutionScene from EVO_MODE_NORMAL and add a condition that checks for EVO_MODE_CANT_STOP as well.

**Comment by kittenchilly** [comment]
> `EVO_LEVEL_ITEM_COUNT_999` sounds good to me too. It makes it clear that the evo method acts if the Player has 999 copies of the specified item to me.

Alright

File: src/party_menu.c (hunk 1 above)
Comment (AlexOn1ine) [reply src/party_menu.c:5788]: Could see if you can rewrite it? So you avoid duplicating code:

```
if (check) 
    BeginEvolutionScene(mon, targetSpecies, TRUE, gPartyMenu.slotId);
else
    BeginEvolutionScene(mon, targetSpecies, FALSE, gPartyMenu.slotId);
```

File: src/pokemon.c [hunk 2]
```diff
@@ -4388,6 +4388,27 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, s
             }
         }
         break;
+    case EVO_MODE_CANT_STOP:
```
Comment (AlexOn1ine) [inline src/pokemon.c:4391]: Doesn't `EVO_MODE_TRADE` do the same thing? So you could rename if to `EVO_MODE_CANT_STOP` and include `EVO_LEVEL_ITEM_COUNT_999` there.

Comment (kittenchilly) [reply src/pokemon.c:4391]: EVO_MODE_TRADE is only called in places that have to do with TradeEvolutionScene, which is not the same thing.

File: src/party_menu.c (hunk 1 above)
Comment (AlexOn1ine) [reply src/party_menu.c:5788]: The code is somewhat confusing right now because of the duplication and I think there is a way to do it with minimal changes.

**Review by AlexOn1ine — CHANGES_REQUESTED** [review]

**Comment by kittenchilly** [comment]
Should be reviewable again.

**Comment by AlexOn1ine** [comment]
Can't use the item
pokeemerald-49

**Comment by kittenchilly** [comment]
> Can't use the item pokeemerald-49

You don't use the item...
You level up while having 999 in your bag.

File: src/pokedex_plus_hgss.c [hunk 3]
```diff
@@ -6653,6 +6654,11 @@ static u8 PrintEvolutionTargetSpeciesAndMethod(u8 taskId, u16 species, u8 depth,
             ConvertIntToDecimalStringN(gStringVar2, evolutions[i].param, STR_CONV_MODE_LEADING_ZEROS, 3);
             StringExpandPlaceholders(gStringVar4, sText_EVO_LEVEL_RECOIL_DAMAGE_FEMALE);
             break;
+        case EVO_LEVEL_ITEM_COUNT_999:
+            item = evolutions[i].param; //item
+            CopyItemName(item, gStringVar2); //item
```
Comment (AlexOn1ine) [inline src/pokedex_plus_hgss.c:6659]: Could you remove both comments?

File: include/constants/pokemon.h [hunk 4]
```diff
@@ -294,14 +294,16 @@
 #define EVO_LEVEL_MOVE_TWENTY_TIMES       47     // Pokémon levels up after having used a move for at least 20 times
 #define EVO_LEVEL_RECOIL_DAMAGE_MALE      48     // Pokémon levels up after having suffered specified amount of non-fainting recoil damage as a male
 #define EVO_LEVEL_RECOIL_DAMAGE_FEMALE    49     // Pokémon levels up after having suffered specified amount of non-fainting recoil damage as a female
+#define EVO_LEVEL_ITEM_COUNT_999             50     // Pokémon levels up after trainer has collected 999 of a specific item
```
Comment (AlexOn1ine) [inline include/constants/pokemon.h:297]: wrong alignment

**Review by AlexOn1ine — COMMENTED** [review]
2 things and then it can be merged :sweat_smile:

**Comment by kittenchilly** [comment]
Addressed the comments

**Review by AlexOn1ine — APPROVED** [review]
